### PR TITLE
Fixes #35547 - Include REX pull in kickstart default

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4and6dhcp.snap.txt [hunk 1]
```diff
@@ -100,6 +100,7 @@ fi
 
 
 
+
```
Comment (ianballou) [inline test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4and6dhcp.snap.txt:103]: Why are there new blank lines here?

**Review by ianballou — COMMENTED** [review]
So I tried to KS a RHEL8 host using this new file with Katello 4.5.  Everything looks fine to me, but anaconda is complaining about an `else` statement on "line 193".  Here's the script that's failing (`/tmp/ks-script-11vxvwee`):

```
exec < /dev/tty3 > /dev/tty3
chvt 3
(
logger "Starting anaconda rhel8.example.com postinstall"




# ens3 interface
real=`grep -l 52:54:00:d0:73:11 /sys/class/net/*/{bonding_slave/perm_hwaddr,address} 2>/dev/null | awk -F '/' '// {print $5}' | head -1`
sanitized_real=`echo $real | sed s/:/_/`

cat << EOF > /etc/sysconfig/network-scripts/ifcfg-$sanitized_real
BOOTPROTO="dhcp"
DOMAIN="example.com"
DEVICE=$real
HWADDR="52:54:00:d0:73:11"
ONBOOT=yes
PEERDNS=yes
PEERROUTES=yes
DEFROUTE=yes
MTU=1500
EOF




echo "Updating system time"
systemctl enable --now chronyd
/usr/bin/chronyc -a makestep
/usr/sbin/hwclock --systohc










  

  echo "##############################################################"
  echo "################# SUBSCRIPTION MANAGER #######################"
  echo "##############################################################"
  echo
  echo "Starting the subscription-manager registration process"

  
    if [ -f /usr/bin/dnf ]; then
      dnf -y install subscription-manager
    else
      yum -t -y install subscription-manager
    fi
  



  

  
    rpm -Uvh http://centos8-luna-demo.foreman-nuc1.example.com/pub/katello-ca-consumer-latest.noarch.rpm
  

  
    subscription-manager register --name="rhel8.example.com" --org='Demo' --activationkey='RHEL 8'
  

  

    

  

  
    
       if [ -f /usr/bin/dnf ]; then
         PACKAGE_MAN="dnf -y"
       else
         PACKAGE_MAN="yum -t -y"
       fi
    

    
      $PACKAGE_MAN install katello-host-tools
    

    
  






# update all the base packages from the updates repository
if [ -f /usr/bin/dnf ]; then
  dnf -y update
else
  yum -t -y update
fi





user_exists=false
getent passwd root >/dev/null 2>&1 && user_exists=true


if $user_exists; then


  mkdir -p ~root/.ssh

  cat << EOF >> ~root/.ssh/authorized_keys
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQCuECETe8EGmP331pUPQBETNJnvADUGaf3pqqcmup9vQ+A6UoHT8cxF2sCAwlDvZYExiKoTxVBg7JgFvFJDNfiuExw7MAIE9bETJwDsAsTC6nHiQ1OAvfkwHiK41jKS4oh6bGqipnJBuK8Ti71K+6xEZRx3IG5eQGz6YP/F1L8OYI/7AYzwiSWUhxEW7flGJZCMDHq25cGG+1lKUnMompAkhd8X3FBXSzqENgpMTUwDY/gtVHZGnjawP+0OM8s7FauTQkG59xZ8+Di//FTG+JZd4q54j6A/wR1KAz0Aa4BbI5AvCHWzlt/njKVMI60IP8p8vP/OitY/Xe/akj47Boev3J+O3abQjlXjMOV4R9ABPaDDpPkB7EZNaulSdPmjGTNpQSPo1pAy7h8v1bOpcareYEcN75Ca7zFQrwNzjyl6N/Hqr4htH88NHZ6Z163DgtGSea0pLvb8yFySVPM6mgWc91OR6ne8/T7WuTn6s2/8YAuq+PLNuu0jltPSl87DiFM= foreman-proxy@centos8-luna-demo.foreman-nuc1.example.com
EOF

  chmod 0700 ~root/.ssh
  chmod 0600 ~root/.ssh/authorized_keys
  chown -R root: ~root/.ssh

  # Restore SELinux context with restorecon, if it's available:
  command -v restorecon && restorecon -RvF ~root/.ssh || true

else
  echo 'The remote_execution_ssh_user does not exist and remote_execution_create_user is not set to true.  remote_execution_ssh_keys snippet will not install keys'
fi


cat << EOF > /root/remote_execution_pull_setup.sh

echo "Starting deployment of REX pull provider"
# check if system is registered
subscription-manager identity > /dev/null
if [ $? -ne 0 ]; then
    echo "Please register with subscription-manager first, then re-run '$0'."
    exit 1
fi

set -e

echo '#'
echo '# Installing packages'
echo '#'

dnf -y install foreman_ygg_worker


echo "Getting configuration from subscription-manager..."
KPPTEMPFILE=$(mktemp kpp_tempfile_XXXXXXXX)
trap "rm -f $KPPTEMPFILE" ERR EXIT
subscription-manager config --list > $KPPTEMPFILE
CONSUMER_CERT_DIR=$(grep 'consumercertdir' $KPPTEMPFILE | cut -d= -f2 | xargs | sed 's/[][]//g')
CERT_FILE=$CONSUMER_CERT_DIR/cert.pem
KEY_FILE=$CONSUMER_CERT_DIR/key.pem
CA_FILE=$(grep 'repo_ca_cert = ' $KPPTEMPFILE | cut -d= -f2 | xargs | cut -d ' ' -f1)
SERVER_NAME=$(grep 'hostname = ' $KPPTEMPFILE | cut -d= -f2 | xargs | cut -d ' ' -f1)

# fail if no server name, cert dir, or ca file
if [ -z "$SERVER_NAME" ] || [ -z "$CONSUMER_CERT_DIR" ] || [ -z "$CA_FILE" ]; then
    echo "Unable to determine config from 'subscription-manager config --list'; exiting"
    exit 1
fi

# fail if client is not registered to a Katello
if ! grep -q 'prefix = \/rhsm' $KPPTEMPFILE; then
    echo "Client is registered to RHSM; exiting"
    exit 1
fi

# set SYSCONFDIR to /etc if it is not set
if [ -z "$SYSCONFDIR" ]; then
    SYSCONFDIR=/etc
fi

# see if /etc/yggdrasil/config.toml exists
CONFIGTOML=$SYSCONFDIR/yggdrasil/config.toml
if [ -f $CONFIGTOML ]; then
    # make a backup of CONFIGTOML
    cp $CONFIGTOML $CONFIGTOML.bak
    cat <<EOF > $CONFIGTOML
# yggdrasil global configuration settings written by katello-pull-transport-migrate
broker = ["mqtts://$SERVER_NAME:1883"]
cert-file = "$CERT_FILE"
key-file = "$KEY_FILE"
ca-root = ["$CA_FILE"]
log-level = "error"
EOF

else
    echo "$SYSCONFDIR/yggdrasil/config.toml not found! Did 'yum install yggdrasil' succeed?"
    exit 1
fi

# start the yggdrasild service
echo "Starting yggdrasild..."
systemctl try-restart yggdrasild
systemctl enable --now yggdrasild

# check status of yggdrasild and fail if it is not running
# possible failure reason: incorrect protocol (should be tcp:// or mqtt://) or port (should be 1883)
# also, cert-file and key-file must be valid
# and broker must be running on the server
yggdrasil status

if ! systemctl is-active --quiet yggdrasild; then
    echo ""
    echo "yggdrasild failed to start! Check configuration in $CONFIGTOML and make sure the broker is running on the server."
    exit $?
fi

echo "Remote execution pull provider successfully configured!"
EOF
chmod +x /root/remote_execution_pull_setup.sh
/root/remote_execution_pull_setup.sh














touch /tmp/foreman_built

chvt 1
) 2>&1 | tee /root/install.post.log
```

Line 193 appears to be:

```
# see if /etc/yggdrasil/config.toml exists
CONFIGTOML=$SYSCONFDIR/yggdrasil/config.toml
if [ -f $CONFIGTOML ]; then
    # make a backup of CONFIGTOML
    cp $CONFIGTOML $CONFIGTOML.bak
    cat <<EOF > $CONFIGTOML
# yggdrasil global configuration settings written by katello-pull-transport-migrate
broker = ["mqtts://$SERVER_NAME:1883"]
cert-file = "$CERT_FILE"
key-file = "$KEY_FILE"
ca-root = ["$CA_FILE"]
log-level = "error"
EOF

else                             <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
    echo "$SYSCONFDIR/yggdrasil/config.toml not found! Did 'yum install yggdrasil' succeed?"
    exit 1
fi
```

However, I pulled out the following code into its own script and verified that it ran, so something else in the KS script must be off...

```
#!/bin/bash

KPPTEMPFILE=$(mktemp kpp_tempfile_XXXXXXXX)
CONSUMER_CERT_DIR=$(grep 'consumercertdir' $KPPTEMPFILE | cut -d= -f2 | xargs | sed 's/[][]//g')
CERT_FILE=$CONSUMER_CERT_DIR/cert.pem
KEY_FILE=$CONSUMER_CERT_DIR/key.pem
CA_FILE=$(grep 'repo_ca_cert = ' $KPPTEMPFILE | cut -d= -f2 | xargs | cut -d ' ' -f1)

if [ -z "$SYSCONFDIR" ]; then
    SYSCONFDIR=/etc
fi

# see if /etc/yggdrasil/config.toml exists
CONFIGTOML=$SYSCONFDIR/yggdrasil/config.toml
if [ -f $CONFIGTOML ]; then
    # make a backup of CONFIGTOML
    cp $CONFIGTOML $CONFIGTOML.bak
    cat <<EOF > $CONFIGTOML
# yggdrasil global configuration settings written by katello-pull-transport-migrate
broker = ["mqtts://$SERVER_NAME:1883"]
cert-file = "$CERT_FILE"
key-file = "$KEY_FILE"
ca-root = ["$CA_FILE"]
log-level = "error"
EOF

else
    echo "$SYSCONFDIR/yggdrasil/config.toml not found! Did 'yum install yggdrasil' succeed?"
    exit 1
fi
```

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4and6dhcp.snap.txt (hunk 1 above)
Comment (adamruzicka) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4and6dhcp.snap.txt:103]: That should be coming from this line https://github.com/theforeman/foreman/pull/9430/files#diff-7755c8fb296da617f75e4c6e39445726b059bc099dde6151bbe5bf468dcb9f20R326 . I could drop that empty line, but then the pull provider setup block would be glued to `<%= snippet "blacklist_kernel_modules" %>`. To me a yet another blank line in the rendered template is worth clarity in the template itself

**Comment by adamruzicka** [comment]
> but anaconda is complaining about an `else` statement on "line 193"

That would be this https://github.com/theforeman/foreman/pull/9426 . save_to_file uses the same heredoc delimiter as something inside the snippet

**Comment by ianballou** [comment]
When the KS script gets around to filling up `remote_execution_pull_setup.sh`, it just hangs forever. I killed the box and started it up to find that  `remote_execution_pull_setup.sh` is empty.  I'm guessing something is still wrong with the file creation snippet.  Here's the related part of the KS script:

```

cat << EOF | sed 's/^ //' > /root/remote_execution_pull_setup.sh
 
 echo "Starting deployment of REX pull provider"
 # check if system is registered
 subscription-manager identity > /dev/null
 if [ $? -ne 0 ]; then
     echo "Please register with subscription-manager first, then re-run '$0'."
     exit 1
 fi
 
 set -e
 
 echo '#'
 echo '# Installing packages'
 echo '#'
 
 dnf -y install foreman_ygg_worker
 
 
 echo "Getting configuration from subscription-manager..."
 KPPTEMPFILE=$(mktemp kpp_tempfile_XXXXXXXX)
 trap "rm -f $KPPTEMPFILE" ERR EXIT
 subscription-manager config --list > $KPPTEMPFILE
 CONSUMER_CERT_DIR=$(grep 'consumercertdir' $KPPTEMPFILE | cut -d= -f2 | xargs | sed 's/[][]//g')
 CERT_FILE=$CONSUMER_CERT_DIR/cert.pem
 KEY_FILE=$CONSUMER_CERT_DIR/key.pem
 CA_FILE=$(grep 'repo_ca_cert = ' $KPPTEMPFILE | cut -d= -f2 | xargs | cut -d ' ' -f1)
 SERVER_NAME=$(grep 'hostname = ' $KPPTEMPFILE | cut -d= -f2 | xargs | cut -d ' ' -f1)
 
 # fail if no server name, cert dir, or ca file
 if [ -z "$SERVER_NAME" ] || [ -z "$CONSUMER_CERT_DIR" ] || [ -z "$CA_FILE" ]; then
     echo "Unable to determine config from 'subscription-manager config --list'; exiting"
     exit 1
 fi
 
 # fail if client is not registered to a Katello
 if ! grep -q 'prefix = \/rhsm' $KPPTEMPFILE; then
     echo "Client is registered to RHSM; exiting"
     exit 1
 fi
 
 # set SYSCONFDIR to /etc if it is not set
 if [ -z "$SYSCONFDIR" ]; then
     SYSCONFDIR=/etc
 fi
 
 # see if /etc/yggdrasil/config.toml exists
 CONFIGTOML=$SYSCONFDIR/yggdrasil/config.toml
 if [ -f $CONFIGTOML ]; then
     # make a backup of CONFIGTOML
     cp $CONFIGTOML $CONFIGTOML.bak
     cat <<EOF > $CONFIGTOML
 # yggdrasil global configuration settings written by katello-pull-transport-migrate
 broker = ["mqtts://$SERVER_NAME:1883"]
 cert-file = "$CERT_FILE"
 key-file = "$KEY_FILE"
 ca-root = ["$CA_FILE"]
 log-level = "error"
 EOF
 
 else
     echo "$SYSCONFDIR/yggdrasil/config.toml not found! Did 'yum install yggdrasil' succeed?"
     exit 1
 fi
 
 # start the yggdrasild service
 echo "Starting yggdrasild..."
 systemctl try-restart yggdrasild
 systemctl enable --now yggdrasild
 
 # check status of yggdrasild and fail if it is not running
 # possible failure reason: incorrect protocol (should be tcp:// or mqtt://) or port (should be 1883)
 # also, cert-file and key-file must be valid
 # and broker must be running on the server
 yggdrasil status
 
 if ! systemctl is-active --quiet yggdrasild; then
     echo ""
     echo "yggdrasild failed to start! Check configuration in $CONFIGTOML and make sure the broker is running on the server."
     exit $?
 fi
 
 echo "Remote execution pull provider successfully configured!"
EOF
chmod +x /root/remote_execution_pull_setup.sh
/root/remote_execution_pull_setup.sh
```

**Comment by ianballou** [comment]
I think the hanging may be related to `KPPTEMPFILE`.  If I copy and paste the entire `cat << EOF | sed 's/^ //' > /root/remote_execution_pull_setup.sh ... EOF` into a host's terminal, it hangs without completion. But, if I preset KPPTEMPFILE to something, everything will run fine (besides the KPPTEMPFILE not being found).

**Comment by adamruzicka** [comment]
This is a gift that keeps on giving. The subshell `$(grep consumercertdir $KPPTEMPFILE ...)` gets evaluated when the script is being written to the disk, at that time, kpptempfile is not set yet so grep reads from stdin.

I don't like this, but how about revisiting https://github.com/theforeman/foreman/pull/9426 and instead of prepending and removing a space we would just encode and later decode the script as base64 or something?

Edit: Like this https://github.com/theforeman/foreman/pull/9436

**Review by ianballou — DISMISSED** [review]
Assuming https://github.com/theforeman/foreman/pull/9436 is the  final `save_to_file` iteration, I'm good with this! I tested it on a CentOS 8 host and it configured yggdrasil perfectly.

File: test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4and6dhcp.snap.txt (hunk 1 above)
Comment (ekohl) [reply test/unit/foreman/renderer/snapshots/ProvisioningTemplate/provision/Kickstart_default.host4and6dhcp.snap.txt:103]: One solution to it is including the empty line inside the `if` block, so:
```erb
<% if plugin_present?('katello') && host_param_true?('enable-remote-execution-pull') -%>
<%= save_to_file('/root/remote_execution_pull_setup.sh', snippet('remote_execution_pull_setup')) %>
chmod +x /root/remote_execution_pull_setup.sh
/root/remote_execution_pull_setup.sh

<% end -%>
<%= snippet "blacklist_kernel_modules" %>
```

And possibly using `<%- end -%>` may also help.

**Comment by stejskalleos** [comment]
Review waiting for https://github.com/theforeman/foreman/pull/9436

**Comment by adamruzicka** [comment]
> Review waiting for https://github.com/theforeman/foreman/pull/9436

It just went in

**Review by ianballou — APPROVED** [review]
I was able to provision a RHEL 8 host and see yggdrasil get setup on it!

**Review by stejskalleos — APPROVED** [review]
:green_apple: **LGTM**

Code looks good, @ianballou confirmed that it works so let's get this in.

**Comment by stejskalleos** [comment]
Thanks @adamruzicka @ianballou !

**Comment by ares** [comment]
Good community demo content :-)

**Pull request merged**